File: deepspace/src/deepspace/Damage.java
package deepspace;
import java.util.ArrayList;
import java.util.Iterator;

public abstract class Damage {
   private int nShields;
   
   Damage(int s){
       nShields=s;
   }

   public abstract Damage copy();

   public abstract Damage adjust(ArrayList<Weapon> w, ArrayList<ShieldBooster> s);

   public abstract void discardWeapon(Weapon w);

   public abstract boolean hasNoEffect();

   public abstract String toString();

    public void discardShieldBooster(){
        if(nShields>0) nShields--;
    }

    public int getNShields(){
        return nShields;
    }

    public abstract DamageToUI getUIversion();

}

File: ruby/Damage.rb
#encoding: utf-8
require_relative 'DamageToUI'
require_relative 'WeaponType'
require_relative 'Weapon'

module Deepspace
  class Damage
      def initialize(nshields)
        @nShields = nshields
      end

      def nShields
        @nShields
      end
      def discardShieldBooster
        if(@nShields > 0)
          @nShields -= 1
        end
      end
      def adjust(s)
        return [s.length, @nShields].min
      end

      def hasNoEffect
        return @nShields==0
      end

      def getUIversion
        return DamageToUI.new(self)
      end
      def to_s
        getUIversion().to_s
      end
  end
end
